modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CustomViewsControls.js

```diff
@@ -194,16 +223,44 @@ const CustomViewsControls = () => {
 					}
 				>
 					<ClayDropDown.ItemList>
-						<ClayDropDown.Item>
-							{Liferay.Language.get('default-view')}
-						</ClayDropDown.Item>
-
 						{Object.keys(customViews).map((id) => (
-							<ClayDropDown.Item key={id}>
+							<ClayDropDown.Item
+								key={id}
+								onClick={() => {
+									viewsDispatch({
+										type:
+											VIEWS_ACTION_TYPES.UPDATE_ACTIVE_CUSTOM_VIEW,
+										value: id,
+									});
+
+									setViewsDropdownActive(false);
+								}}
```

Comment:
This is similar to what @diegonvs said in a previous comment, but there's multiple functions declared inside this component (`getNextCustomViewId`, `saveCustomView`, etc) and a lot of inline `onClick` handlers. These get re-created with every render, forcing the component that receives them as a prop to re-render even if there's no actual change.

We could wrap them in `useCallback` so that they only recalculate when necessary.

Reply:
@tinycarol @diegonvs  Thanks for reviews! :heart: I think all three comments boil down to handling internal assignments, so I'll tackle it here:

> there's multiple functions declared inside this component (getNextCustomViewId, saveCustomView, etc) and a lot of inline onClick handlers. These get re-created with every render, forcing the component that receives them as a prop to re-render even if there's no actual change.

I am not sure that this should be a concern. Maybe I'm missing something:
- When the whole app renders for the first time, there are no _additional_ renders caused by functions declared in the component. There is no state change in the assignment of an arrow function to a variable (`const getNextCustomViewId = () => {...}`). There is no execution of code within function, that may change state.
- When a component re-renders on a change of state declared in that component, the _whole_ component re-renders, all child components do. It doesn't matter if the props of child components changed or not. For example, when `viewsDropdownActive` changes on UI click, the whole `CustomViewsControls` re-renders. If you put a component with no props as a sibling of `<ManagementToolbar.Item>`s, it will still re-render. As far as I know, in React there is no way to update only part of a component. The way to indirectly do this is to refactor component into several sibling components, and move state down, into one of them.

The only downside of having internal declarations is that we have a few (six) unnecessary assignments with each render. But, this a very small hit on performance. It would be a different case if we had code that executed on each render. Something like [this](https://github.com/liferay/liferay-portal/blob/master/modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/components/rules/RuleEditorModal.js#L78-L87). This is where `useMemo` shines. 

The benefit of internal declarations is that we can reuse scope variables, reduce boilerplate of props propagation, ease maintenance, and have a clearer hierarchy. Internal helper functions and components are literally internal to main component. Having said that, `useCallback` is a good suggestion! With it, a small hit on performance becomes even smaller. The cost is a bit more code, but with other benefits preserved. It seems like a reasonable compromise.

The same logic applies to internal components, except I don't think we can use `useCallback` with them. Even if we can, we probably shouldn't, it would be very hacky.

Reply:
Thank you so much for the thorough explanation, I've been using this wrong for years! TIL that `useCallback` is worthless if the components that receive those callbacks aren't memoized. I remember reading `components render only when their props or state change`, so I thought that when a component renders, its children won't re-render if the props you pass down are referentially the same.

![shame](https://media3.giphy.com/media/eP1fobjusSbu/giphy.gif?cid=ecf05e47q7m8ebt5tn6z02lz01diy9oai3rntby1btw2wj1o&rid=giphy.gif&ct=g)

BRB, gonna remove React from my CV.